### PR TITLE
Jira-311: fixing phage category filter and other category-related bugs

### PR DESCRIPTION
**Pull request by jameshowardwang**
# Description

Phage category filter wasn't working properly, in fact doing the opposite of what it should (when Phage was selected, then it would be excluded). Also, whenever any category was removed using the 'X' on the blue pill, phage would be removed. Finally, whenever any category was removed using the 'X' or filters were reset, the state of the MultipleDropdown would silently preserve the old state, which would show up in weird ways the next time an item was selected.

All three of those issues are resolved in this diff.

## Type of change

- [ x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix a feature that knowingly causes existing functionality to not work as expected)

# Test and Reproduce
Outline the steps to test or reproduce the PR here.

1. Go to http://localhost:3000/samples/27?pipeline_version=1.0&background_id=4
1. Change category filter to any subset of {Virus, Phage, Viroids, Uncategorized}
1. Click 'X' on any of the blue category pills
1. Hit 'Clear All' or "Show All" in the text filter
1. Hit Read Specificity: "All" / "Specific Only"

List of results should match expected behavior (and compared with results in db)

## Impacted Areas in Application
List general components of the application that this PR will affect:

Which pages?

- [ ] Login Screen
- [ ] Single Upload Page
- [ ] Batch Upload Page
- [ ] All Projects Page
- [ ] Sample Details Page
- [x] Report Page

Specific components 

* MultipleDropdown / filter tags / Report page

# Testing Script:

###Login Page
* Successfully login to the application and get taken to the project page

###Single Upload
* Using an AWS path that doesn't exist throws an error
* Succesfully uploaded files from a folder (use test host)
* Uploading a duplicate file throws an error
* Pipeline gets initiated successfully 

###Batch Upload
* Using an AWS path that doesn't exist throws an error
* Succesfully uploaded files from a folder (use test host)
* Navigate to batch upload settings page after upload 
* Host selection from previous page still selected 
* Uploading a duplicate file throws a warning
* Pipeline gets initiated successfully 

###All Project Page
* Switch between projects 
* Search for a sample
* Navigate through different pages of samples
* Select Upload new sample button
* Select sample to see generated report

###Sample Details Page
* Information is consistent from Project Page
* Notes are editable and save successfully 
* Can navigate to report page 

###Report Page
* Data loads succesfully 
* Collapse and Uncollapse button works
* Clicking on the taxonomy name gives you total reads for that taxon
* Category filter works
* Genus Search works
* Filters work as expected
* Disabling Filters works
* Filtering by column works 

# Checklist:

- [ ] I have run through the testing script to make sure current functionality is unchanged
- [x] I have done relevant tests that prove my fix is effective or that my feature works
- [x] I have spent time testing out edge cases for my feature
- [ ] I have updated the test script or pull request template if necessary
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/assets/src/components/PipelineSampleReport.jsx [hunk 1]
```diff
@@ -333,16 +335,13 @@ class PipelineSampleReport extends React.Component {
 
       for (var i = 0; i < thresholded_taxons.length; i++) {
         let taxon = thresholded_taxons[i];
-
-        if (includedCategories.indexOf(taxon.category_name) >= 0) {
-          // In the included categories
-          selected_taxons.push(taxon);
-        } else if (
+        if (
+          includedCategories.indexOf(taxon.category_name) >= 0 ||
```
Comment (cdebourcy) [inline app/assets/src/components/PipelineSampleReport.jsx]: perhaps we could deduplicate lines 339-342 and 357-360 by defining a helper function called `isIncluded(taxon, includedCategories, includedSubcategories)`, which should also help with readability. Particularly if you also move lines 330-334 in there

File: app/assets/src/components/ui/controls/dropdowns/MultipleDropdown.jsx [hunk 2]
```diff
@@ -25,6 +25,12 @@ class MultipleDropdown extends React.Component {
     this.passedOnChange && this.passedOnChange(e, value);
   }
 
+  componentWillReceiveProps(nextProps) {
```
Comment (cdebourcy) [inline app/assets/src/components/ui/controls/dropdowns/MultipleDropdown.jsx]: https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops

Comment (tfrcarvalho) [reply app/assets/src/components/ui/controls/dropdowns/MultipleDropdown.jsx]: This is not available on the current react version we are using.
Created a task for that and necessary refactoring: https://jira.czi.team/browse/IDSEQ-325
